### DISCUSSION TITLE
Null Handling

### DISCUSSION THREAD

**Discussion by zeuchste** [nebulastream/nebulastream#45] [Ideas]
We do not handle/support NULL values.

From the discussion in _Originally posted by @keyseven123 in https://github.com/nebulastream/nebulastream-public/discussions/4#discussioncomment-9654084_ we concluded that Null values are not a topic in the node engine but rather an issue for the compiler

There are some open questions:
- Do we use NULL as a default value for every data type?
- Do we need to integrate a code check for NULL to every access to a value, thus creating code like `if(a != NULL) do; else use it`
-- We need to find out where do we actually need this for all accesses or only for particular accesses?
-- Is the handling of NULL values operator dependent or now?
-- How are arithmetic functions are actually defined on nullable types
-- Are there actually null and not-null datatypes? and if yes how do they behave to each other
- Do we want to support NULLable as a property for each attribute in the schema and if not how do we handle then NULL values
- What would be the outcome of NULL+ 3? Is it 3, NULL, or do we raise an error, or is it undefined behavior and we pretend this case would never happen.
- What about dividing by null?
- For window aggregations: what if one value is null? Is the aggregation result also null?
- On the logical level we can just append a nullable flag to the data type. And on the physical memory layout level we could consider a bitfield in front of every tuple, or maybe even group multiple fields to not totally mess up the alignment. The nth bit would represent the absence/presence of the nth nullable field. Is there any other possible solution?
- One important aspect is also that expressions could produce null values. ParseInt("Hello")

Those question arise from the discussions. A possible next step should be how other compilation-based systems handle this or what are well established solutions for that?

In addition we agreed that this item is blocked by #25 as we need first to include nautilus as a lib before making those changes

**Comment by PhilippGrulich** [comment]
I don't get that null handling is only something of the optimizer? I mean the functions and operator need to check for it?

We can just follow sql semantics, this makes these cases all clear I think.

**Reply by adrianmichalke** [reply]
> We can just follow sql semantics, this makes these cases all clear I think.

Yes, I think we should stick to SQL wherever its possible. Do you have a preferred reference for sql semantics?

[Here](https://community.snowflake.com/s/article/NULL-handling-in-Snowflake) are some definitions by Snowflake: Expressions with NULL return NULL. In aggregation NULL is evaluated as 0 or dismissed.

Flink has defined similar semantics for NULL [here](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/functions/systemfunctions/). Except: Expressions like x > NULL evaluate to UNKNOWN. I could not find in their docu how they handle UNKNOWN. NULL and UNKNOWN are the same according to [SQL ANSI](https://webdocs.cs.ualberta.ca/~yuan/courses/db_readings/ansi-iso-9075-2-1999.pdf).

**Reply by PhilippGrulich** [reply]
I think this sounds good. I think we can implement it using functions.

**Reply by zeuchste** [reply]
The question that remains is where do we actually need to do something for that. My current understanding is that null handling would mainly require compiler changes like:

`if(x == NULL)
then do special things;
else
do logic`

Other cases might be
1.  is NULL handling in addition a operator thing such that each operator like an SUM needs to check for NULL and do different things based on NULL values
2. Is null handling and issue with serialization so do we need to changes here
3. Is there anything we need to do in the Optimizer like checking nullable and avoid for example expressions of a + b if b is nullable. 
4. Is null handling something that we need to express in the API and/or schema definition

I do see potential changes at all those levels.

**Reply by PhilippGrulich** [reply]
I would think all four points are true and need to addressed. 
Function/operators need to check null and set the flag on the output. We have to represent null flags in the tuple layout. The optimizer should rewrite queries where it is sure that fields are not nullable and the user should be able to set not null as a field property.
